src/Microsoft.OData.Core/Uri/ODataUriSlim.cs

```diff
@@ -0,0 +1,139 @@
+﻿// <copyright file="ODataUriSlim.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using System.Diagnostics;
+using System.Collections.Generic;
+using Microsoft.OData.UriParser.Aggregation;
+using Microsoft.OData.UriParser;
+
+namespace Microsoft.OData
+{
+    /// <summary>
+    /// A lightweight version of <see cref="ODataUri"/> that is
+    /// used by the writer to efficiently create a copy of the URI when entering
+    /// each scope. This is used to avoid cloning <see cref="ODataUri"/>
+    /// excessively.
+    /// </summary>
+    internal struct ODataUriSlim
+    {
+        private readonly ODataUri odataUri;
+
+        /// <summary>
+        /// Creates an instance of <see cref="ODataUriSlim"/>
+        /// by copying the provided <see cref="ODataUri"/> instance.
+        /// </summary>
+        /// <param name="odataUri">The URI to copy.</param>
+        public ODataUriSlim(ODataUri odataUri)
+        {
+            Debug.Assert(odataUri != null, "odataUri != null");
+            this.odataUri = odataUri;
+            this.SelectAndExpand = odataUri.SelectAndExpand;
+            this.Path = odataUri.Path;
+        }
+
+        /// <summary>
+        /// Creates an instance of <see cref="ODataUriSlim"/> by
+        /// cloning the provided <see cref="ODataUriSlim"/>.
+        /// </summary>
+        /// <param name="odataUriSlim">The URI to copy.</param>
+        public ODataUriSlim(ODataUriSlim odataUriSlim)
+        {
+            this.odataUri = odataUriSlim.odataUri;

```

Comment:
In quite a few places, the `ODataUriSlim` property is declared as nullable. We have any guarantees that this constructor can't be called with a null value for `odataUriSlim`?

Reply:
Yes, it's guaranteed by the compiler. You cannot assign `null` to `ODataUriSlim` variable. Similarly, you cannot assign a variable of type `ODataUriSlim?` (nullable) to a variable of type `ODataUriSlim` (not nullable). You can only go from `ODataUriSlim` to `ODataUriSlim?`, but not the other way around. Otherwise, the code will not compile.